airflow/contrib/hooks/bigquery_hook.py

```diff
@@ -495,6 +495,77 @@ def get_tabledata(self, dataset_id, table_id,
             .execute()
         )
 
+    def run_table_upsert(self, dataset_id,  table_resource):
+        """
+        creates a new, empty table in the dataset;
+        If the table already exists, update the existing table.
+        Since BigQuery does not natively allow table upserts, this is not an atomic operation.
+        :param dataset_id: the dataset to upsert the table into.
+        :type dataset_id: str
+        :param table_resource: a table resource. see https://cloud.google.com/bigquery/docs/reference/v2/tables#resource
+        :type table_resource: dict
+        :return:
+        """
+        # check to see if the table exists
+        table_id = table_resource['tableReference']['tableId']
+        table_exists = False
+        tables_list_resp = self.service.tables().list(projectId=self.project_id, datasetId=dataset_id).execute()
```

Comment:
Wonder if project_id should be a param, and default to self.project_id? Reason I ask is that I wonder if this has similar problems to the other job APIs, where you might want to insert a table in shared dataset on a project that you don't have view access to. In such a case, it would be impossible to use this. What you'd want to say is use the hook for project-a, but insert the table on project-b.